### PR TITLE
Support for XNNPACK max pooling operator.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35896 Add XNNPACK Clamp.
* #35791 Enable stateless XNNPACK linear.
* #35790 Enable stateless XNNPACK convolutions.
* **#35354 Support for XNNPACK max pooling operator.**
* #35351 Propagate input tensor names in XNNPACK backend.

Differential Revision: [D20821862](https://our.internmc.facebook.com/intern/diff/D20821862)

### CODE REVIEW

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+  // Normalize the parameters.
+  const internal::pooling::Parameters parameters{
+    kernel_,
+    padding_,
+    stride_,
+    dilation_,
+  };
+
+  return xnnpack::internal::available() &&
```
Comment (ailzhang) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp]: This condition is pretty hard to read, it'll be great if you can add a verbose comment to describe when xnnpack is used.

Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:60]: Done, thanks.

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  // Make sure we are not dealing with an unorthodox configuration.
+  if (kernel_.empty() || padding_.empty() || dilation_.empty()) {
+    return false;
+  }
+
+  // Stride can be legitimately empty, in which case it is to be defaulted to kernel size.
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp:34]: Is this right? Default to kernel size or to 1? I dont know the convention but just making sure.

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+      (4 == input.dim()) &&
+      (c10::DeviceType::CPU == input.device().type()) &&
+      (kFloat == input.scalar_type()) &&
+      !input.requires_grad() &&
+      // Kernel
+      (2 == parameters.kernel.size()) &&
+      (parameters.kernel[Layout::Parameter::height] > 0) &&
+      (parameters.kernel[Layout::Parameter::width] > 0) &&
+      ((parameters.kernel[Layout::Parameter::height] *
+        parameters.kernel[Layout::Parameter::width]) > 1) &&
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp:71]: Why this condition? If kernel high and width both must be integers and > 0, then this condition is implied, no?

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      // Dilation
+      (2 == parameters.dilation.size()) &&
+      (parameters.dilation[Layout::Parameter::height] > 0) &&
+      (parameters.dilation[Layout::Parameter::width] > 0) &&
+      // Ceil Mode
+      !ceil_mode &&
+      // Output Min / Max
+      (output_max > output_min) &&
+      // Output
+      (pooling_output_shape(
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp:89]: Does this function assume parameters such as kernel, padding, stride both must be symmetrical, meaning both height and width of these parameters be the same? Because we are not enforcing that anywhere else.

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    const IntArrayRef kernel_,
+    const IntArrayRef padding_,
+    IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  // A call to max_pool2d must have been gated by a call to use_maxpool2d, so
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp:117]: Can this be enforced somehow? That is, if you come here without going through use_maxpool2d, then it is caught? And if it is not possible then it would be good to couple these two together such that this is not possible. I guess the trouble would be the API of max_pool2d. Maybe it can return optional Tensor?

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp (hunk 2 above)
Comment (dreiss) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:34]: Pytorch defaults to 1 for convolution, but kernel_size for pooling.  This matches how conv and pool are "normally" used.

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:71]: 1x1 max pooling is not meaningful and XNNPACK prohibits that use case.  As for the check, let's read > 0 as >= 1 to make it more intuitive.  With that in mind, for any integer M and N greater than or equal to 1, MxN, 1xN, and Mx1 are all valid sizes except for 1x1 which is what this check is guarding against.

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp (hunk 5 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:117]: I cannot think of a way since both these functions are pure and stateless.

Comment (dreiss) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:117]: I think the idea is that you would remove `use_max_pool2d` as a separate function and bring all that logic into `max_pool2d`.  That function would first perform the checks.  If they fail, it would return `nullopt`, and if they succeed, it would perform the max pool and return the result tensor.  Then the calling code would change from

```
if (can_use_xnnpack()) {
  return use_xnnpack();
}
```
to
```
optional<Tensor> ret = use_xnnpack();
if (ret) {
  return *ret;
}
```

It's not clear to me whether this would be better or worse than the current implementation.

File: aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp [hunk 6]
```diff
@@ -15,6 +15,7 @@ using internal::linear::createLinearClampPrePackOpContext;
 using internal::convolution2d::createConv2dClampPrePackOpContext;
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:15]: Looks like no change to this file.  Want to revert it?

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    kernel_,
+    padding_,
+    stride_,
+    dilation_,
+  };
+
+  // Here are the list of conditions required for this code path to be taken:
+  // * Input must be 4D CPU float tensor with no gradients.
+  // * Kernel must be a 2D IntArrayRef containing two positive numbers.
+  //   Furthermore, 1x1 kernels are neither valid nor meaningful.
```
[the remark this answers is not in the document]
Comment (dreiss) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:50]: I see you commented below that XNNPACK prohibits this.  That's fine, but I'd update this comment to explain that that is the reason for returning false in this case.

**Review by dreiss — APPROVED** [review]

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp:50]: Right, XNNPACK errors on this input.  Will improve the expand the comment with the explanation.

File: aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp (hunk 6 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:15]: Will do.
